tests/TypeScriptTest.sh

```diff
@@ -14,15 +14,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-pushd "$(dirname $0)" >/dev/null
+set -e
 
 # clean node_modules to make sure we depend on latest local flatbuffers at ../
 rm -rf node_modules
 npm install
 
-../flatc --ts --gen-name-strings --gen-mutable --gen-object-api -o ts -I include_test monster_test.fbs
-../flatc --gen-object-api -b -I include_test monster_test.fbs unicode_test.json
-../flatc --ts --gen-name-strings --gen-object-api -o ts union_vector/union_vector.fbs
+if [ -x ../flatc ]; then
```

Comment:
I don't think running tests with missing flatc makes much sense - it should rather fail immediately.

Reply:
The rationale is that we have the artifacts checked in so it's not really a requirement to run the tests, but it's practical when doing local development.

Reply:
When doing local dev, I don't really see a situation in which someone doesn't build flatc, but wants to run tests. What would be the purpose, since flatc is the thing being worked on?

Reply:
The argument for not requiring flatc is for the automated CI test run.

Reply:
Theoretically it can also be useful to be able to do development on https://github.com/google/flatbuffers/tree/master/ts without having to build and regenerate the code gen, but that's probably not important.

Reply:
Then you're not really testing changes, but rather re-testing some previous build.

Reply:
We have a dedicated independent ci step that builds flatc and generates all code gen and make sure it is not diff from the commited.

Reply:
Ok, I see two possible situations:

1. That dedicated step builds flatc, thus making this check unneeded.
2. Nothing builds flatc, thus making tests unneeded at all, since there's no way to test actual changes.

In both cases, that if seems to be unnecessary. I might be missing a use case here, but it simply looks strange at the moment.

Reply:
I'm not sure we understand each other here. I'll try to be more clear:

1. We have steps that build flatc and run `check-generate-code.sh` and `check-grpc-generated-code.sh`. These steps ensure two things - that we have the up to date generated code for all languages in commited into the repo and that it is identical with both GCC and clang.
2. We have steps per language to build (if needed) and run tests - they do not need to build flatc and generate code again because we already have a step that verifies the commited code gen. This step was missing for TS so we did not run the testcases at CI for TS, that is what this PR is about.
3. I only made the conditional to avoid not found messages at CI build log since that might be interpreted as something is wrong. I did not want to remove the flatc lines from the test runner because it is useful for local development.

Reply:
My issue is with point 3 - I see no use case for this. Either we build flatc and test new code, or we don't build flatc and don't run tests, since there's nothing to test at that point. In both cases, checking for flatc existence doesn't seem to have a purpose - first situation already ensures it exists, while the second should not run those tests at all.

Reply:
Yes the conditionals have no relevance for point 3, as I wrote I added them only to avoid error log messages for point 2. I don't understand why you are saying we should not run tests in point 2 which is the only purpose of point 2?! If we remove point 2 we are back to what we have in master now which is that we do not run the tests for CI (for TS only, all other langs do have a test run step in place!) and we've had several commits that was merged with undetected failing tests because of that.

Reply:
To shortcut the discussion which seem to be leading nowhere :) Will you accept this if I simply remove the conditionals? (they where *only* added to make CI log more clean)

Reply:
I'm not saying we should stop running test in point 2. My reasoning is: if we don't build flatc, then there's no point in running tests, since there's nothing new to test. That's why tests require flatc to exist, otherwise they don't test actual changes made. The check is therefore unnecessary.

Also, if building flatc silently failed for any reason, the tests should also fail. This check makes them pass in such situations.

Reply:
And again my argument is we have already made certain that we have the exact same flatc code gen commited as would be generated. And yes that means we're not testing anything new but it ensures we enforce test runs on what is commited. That is what CI is for, run the same thing that is expected of the local developer but that might be forgotten or passing due to some difference in what is locally.

Reply:
If flatc isn't built, then something went silently wrong in the previous CI step, therefore we should immediately fail. I consider this a broken invariant, that's why I'm not happy with us treating it as a normal situation.

Reply:
The CI steps are independent, can run concurrently and it's a good thing CI run time is as short as possible. Making CI steps dependent on each other increases complexity and CI run time. I really can't see the broken invariant as we ensure commited code gen is valid and exactly the same as would be produced by flatc. In fact if we code gen yet again in each test step and for whatever strange reason the resulting code gen is different from what is commited and test passes on that it's a false positive.

Reply:
We cannot treat those steps as independent - tests have to depend on building, otherwise they might be run before the building process. After all, we're testing the effects of the new build, not a collection of test files from an arbitrary point in time. If the build fails, we cannot test it => invariant is broken.

Reply:
But that means you consider the whole CI (at GitHub actions) setup wrong and I must disagree. CI passes only if all steps pass so naturally if we cannot build flatc the CI will not pass as a whole. And we do have a step that compiles flatc and explicitly verifies that the resulting code gen is *exactly* the same as is commited. If we did not commit code gen we would definitely have to do something else, but we commit the code gen for several good reasons.

Reply:
Notice that we also have a dependency between tests and build, although an implicit one - tests assume some step verifies committed test data comes from the current build. Explicitly stated or not, this is a kind of dependency. If the build output verification step disappears or silently stops working, tests will give wrong result, because their assumption (=dependency) gets broken. They will simply happily test old data.

Don't get me wrong - I'm not blocking merging this PR, but I think we have a problem with our CI configuration if we not only allow implicit dependency on build verification, but we also explicitly allow build to not exist.

Reply:
Ok I see your point about that the implicit dependency can be seen as too weak or unreliable, though I'm not sure I see it as a big issue.

Anyway I suggest merging this as it is better than nothing and cleans up the duplicate and confusing artifacts we have now.

Reply:
I am fine either way, but if it needs a tie breaker, I would agree that I prefer whatever setup fails harder/earlier whenever some part is missing, which would suggest @krojew's approach. But again it seems to be a minor difference in this case.